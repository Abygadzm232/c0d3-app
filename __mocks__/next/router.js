global.window = Object.create(window)
Object.defineProperty(global.window, 'location', {
  value: { pathname: '/not-root' } // make sure pathname isnt '/' by default
})
const push = jest.fn(path => (global.window.location['pathname'] = path))
export const useRouter = jest.fn().mockImplementation(() => {
  return {
    route: '/',
    pathname: '/',
    query: {},
    asPath: '/',
    push: push,
    replace: async () => true,
    reload: () => null,
    back: () => null
  }
})
