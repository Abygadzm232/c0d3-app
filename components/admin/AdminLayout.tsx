import React from 'react'
import LoadingSpinner from '../../components/LoadingSpinner'
import Layout from '../../components/Layout'
import _ from 'lodash'
import { GetAppProps } from '../../graphql'

export const AdminLayout: React.FC<GetAppProps> = ({ data, children }) => {
  const { loading, error, session } = data

  if (loading) {
    return <LoadingSpinner />
  }

  if (error) {
    return (
      <Layout>
        <h1>Error</h1>
      </Layout>
    )
  }

  if (!session) window.location.href = '/login'

  const isAdmin = _.get(session, 'user.isAdmin', false)

  if (isAdmin !== 'true') {
    return (
      <Layout>
        <h1>You must be admin to access this page</h1>
      </Layout>
    )
  }

  return <Layout>{children}</Layout>
}
