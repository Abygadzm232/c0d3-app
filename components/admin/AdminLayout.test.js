import React from 'react'
import { render, waitFor } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import { AdminLayout } from './AdminLayout'
import { useRouter } from 'next/router'
jest.mock('next/router')

// Mock global.window
global.window = Object.create(window)
Object.defineProperty(global.window, 'location', {
  value: { pathname: '/not-root' } // make sure pathname isnt '/' by default
})
useRouter.mockReturnValue({
  push: jest
    .fn()
    .mockImplementation(path => (global.window.location.pathname = path)),
  asPath: '/'
})
describe('AdminLayout test', () => {
  test('Should return loading spinner when loading', async () => {
    const { findByRole } = render(
      <MockedProvider addTypename={false}>
        <AdminLayout data={{ loading: true }} />
      </MockedProvider>
    )
    const element = await findByRole('heading', { name: /loading/i })
    expect(element).toBeTruthy()
  })
  test('Should return Error on error', async () => {
    const { findByText } = render(
      <MockedProvider addTypename={false}>
        <AdminLayout data={{ error: true }} />
      </MockedProvider>
    )
    const element = await findByText(/Internal server error/i)
    expect(element).toBeTruthy()
  })
  test('Should redirect to login page if no session', async () => {
    render(
      <MockedProvider addTypename={false}>
        <AdminLayout data={{ data: {} }} />
      </MockedProvider>
    )
    await waitFor(() =>
      expect(global.window.location.pathname).toEqual('/login')
    )
  })
  test('Should render children', async () => {
    const { container } = render(
      <MockedProvider addTypename={false}>
        <AdminLayout
          data={{
            session: {
              user: {
                isAdmin: 'true'
              }
            }
          }}
        >
          <h1>Test title</h1>
        </AdminLayout>
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
  })
})
