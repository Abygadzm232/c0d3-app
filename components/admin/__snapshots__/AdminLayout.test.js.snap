// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminLayout test Should render children 1`] = `
<div>
  <div
    class="container-fluid bg-primary d-flex flex-column justify-content-center align-items-center"
    style="min-height: 100vh; position: fixed; top: 0px; left: 0px; z-index: 1;"
  >
    <h1
      class="text-white font-weight-bold display-3"
      style="font-family: 'PT Mono', sans-serif;"
    >
      C0D3
    </h1>
    <div
      class="spinner-border text-white"
      role="status"
    >
      <h1
        class="sr-only"
      >
        Loading...
      </h1>
    </div>
  </div>
  <div
    class="container"
  >
    <h1>
      Test title
    </h1>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    Â© Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;
